lib/register_common/services/bulk_transformer.rb

```diff
@@ -0,0 +1,56 @@
+require 'register_common/services/file_reader'
+
+require_relative 'msg_handler'
+
+module RegisterCommon
+  module Services
+    class BulkTransformer
+      BATCH_SIZE = 25
+      DEFAULT_PARALLEL_FILES = 1
+      NAMESPACE = 'BULK_TRANSFORMER'.freeze
+
+      def initialize(s3_adapter:, s3_bucket:, set_client:, file_reader: nil, batch_size: nil)
+        @s3_adapter = s3_adapter
+        @s3_bucket = s3_bucket
+        @set_client = set_client
+        @file_reader = file_reader || RegisterCommon::Services::FileReader.new(
+          s3_adapter:,
+          batch_size: (batch_size || BATCH_SIZE),
+        )
+        @msg_handler = MsgHandler.new(s3_adapter:, s3_bucket:)
+      end
+
+      def call(s3_prefix, parallel_files: nil, &block)
+        processed_files = set_client.init_set("#{NAMESPACE}:#{s3_prefix}")
+
+        s3_paths = s3_adapter.list_objects(s3_bucket:, s3_prefix:)
+
+        s3_paths.each_slice(parallel_files || DEFAULT_PARALLEL_FILES) do |s3_paths_batch|
+          threads = []
+
+          s3_paths_batch.each do |s3_path|
+            next if processed_files.contains?(s3_path)
+
+            threads << Thread.new do
```

Comment:
Are we sure using `set_client` is thread-safe, here? `Redis.new` is used in `RedisAdapter`, which isn't wrapped in a connection pool…

Reply:
Discussed. It's not clear whether it's thread-safe, but it's apparently a refactor of an existing approach which also might not have been thread-safe (or might have been), so this can be considered out-of-scope since it isn't causing any issues.